Framework/Geometry/test/BeamlineRayTracerTest.h

```diff
@@ -0,0 +1,269 @@
+#ifndef BEAMLINERAYTRACERTEST_H_
+#define BEAMLINERAYTRACERTEST_H_
+
+#include "MantidGeometry/Instrument/ComponentInfo.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+#include "MantidGeometry/Instrument/InstrumentVisitor.h"
+#include "MantidGeometry/Instrument/RectangularDetector.h"
+#include "MantidGeometry/Objects/BeamlineRayTracer.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidTestHelpers/ComponentCreationHelper.h"
+
+#include <boost/make_shared.hpp>
+#include <cxxtest/TestSuite.h>
+
+using namespace Mantid::Geometry;
+using Mantid::Kernel::V3D;
+using namespace ComponentCreationHelper;
+namespace RayTracer = Mantid::Geometry::BeamlineRayTracer;
+using Links = Track::LType;
+
+class BeamlineRayTracerTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static BeamlineRayTracerTest *createSuite() {
+    return new BeamlineRayTracerTest();
+  }
+  static void destroySuite(BeamlineRayTracerTest *suite) { delete suite; }
+
+  BeamlineRayTracerTest() {
+    // Start logging framework
+    Mantid::Kernel::ConfigService::Instance();
+  }
+
+  void
+  test_that_a_trace_for_a_ray_that_intersects_many_components_gives_these_components_as_a_result() {
+    // Create the test objects to use
+    create_instrument_and_componentInfo();
+
+    // Vector for a ray
+    V3D testDir(0., 0., 1);
+
+    // Do a trace and store the results
+    Links results = RayTracer::traceFromSource(testDir, *m_compInfo);
+
+    // Check size
+    TS_ASSERT_EQUALS(results.size(), 2);
+
+    // Check they are actually what we expect: 1 with the sample and 1 with the
+    // central detector
+    IComponent_const_sptr centralPixel =
+        m_testInstrument->getComponentByName("pixel-(0;0)");
+    IComponent_const_sptr sampleComp = m_testInstrument->getSample();
+
+    if (!sampleComp) {
+      TS_FAIL("Test instrument has been changed, the sample has been removed. "
+              "Ray tracing tests need to be updated.");
+      return;
+    }
+    if (!centralPixel) {
+      TS_FAIL("Test instrument has been changed, the instrument config has "
+              "changed. Ray tracing tests need to be updated.");
+      return;
+    }
+
+    // Iterate through the results
+    Links::const_iterator resultItr = results.begin();
+    Link firstIntersect = *resultItr;
+
+    // Checks
```

Comment:
Would be useful to have some commented context to these tests. How would someone in future debug this.. you could help them by explaining why they should find these links.